### PR TITLE
[FLINK-5520] [table] Disable outer joins with non-equality predicates

### PR DESCRIPTION
**Pull request by lincoln-lil**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by wuchong** [comment]
Hi @lincoln-lil , thank your for the PR. I think the issue is only for 1.2.0, so please create a new pull request to commit into `release-1.2`.  For `master` we will directly fix the bug. 

I have taken a quick look at it. I think a better approach is to disable outer joins with non-equality in [`DataSetJoinRule`](https://github.com/apache/flink/blob/master/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/dataSet/DataSetJoinRule.scala#L43),  not mix the code in SQL translation. such as 

```scala
override def matches(call: RelOptRuleCall): Boolean = {
    val join: LogicalJoin = call.rel(0).asInstanceOf[LogicalJoin]

    val joinInfo = join.analyzeCondition

    // joins require an equi-condition or a conjunctive predicate with at least one equi-condition
    // and outer joins with non-equality predicates is not supported currently
    !joinInfo.pairs().isEmpty && (join.getJoinType == JoinRelType.INNER || joinInfo.isEqui)
  }
```

 And also please add checks in [`operators.scala`](https://github.com/apache/flink/blob/release-1.2/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/logical/operators.scala#L455) to give a good exception for TableAPI outer joins.

**Comment by lincoln-lil** [comment]
Hi, wuchong, thanks for your review. I'll switch the pr based on 1.2.0.
I think it's not necessary doing validation repeatedly both in DataSetJoinRule.java and operators.scala, and there already being several validations in DataSetJoin.translateToPlan function, why don't we put the same validation together to reduce code?

**Comment by shaoxuan-wang** [comment]
@wuchong @lincoln-lil , I do not think this is just for 1.2.0. The out join with non equality functionality is broken on master. We should deliver this fix to the master as well as release 1.2.0 until we have concrete plans to fix non equality functionality.

**Comment by fhueske** [comment]
I agree with @lincoln-lil and @shaoxuan-wang, this fix should go into the `1.1-release`, `1.2-release`, and `master` branches. We will later add support for non-equality predicate on left/right outer joins to master and include it for 1.3.0.

Regarding the check, I agree with @wuchong. We implement these checks in  `RelOptRule.matches()` to prevent that an invalid plan is constructed. Otherwise it may happen that we end up with an invalid plan although, there would be a valid (but more expensive) alternative. We have additional checks in the DataSet operators to be on the safe side but these should never be triggered. If this happens, something is wrong with the optimization phase. So its kind of a safety net. Although, I don't think there are rules to generate alternative plans to execute an outer join with non-equality predicate, I would like to keep the pattern for consistency reasons. 
Also the check in the validation phase of the Table API is useful, because the validation happens immediately when a function on a `Table` is called. In contrast, the check on `DataSetJoin` would happen when the translation is triggered, i.e., when the program is optimized and translated into a DataSet program. Throwing an exception early is helpful for users because they can much easier identify an invalid statement.

**Comment by lincoln-lil** [comment]
@fhueske agree with you that the check in validation phase of the Table API is useful for failing fast( IMO, it seems no difference on user experience that when an exception be thrown either in validation or translation phase ?), but there may be other rules can match and generate alternative plans to execute an outer join. For instance, we internal implemented a HbaseTableSource extends TableSource that can be joined as a TableSourceTable. We created a new rule that is used to match inner and outer joins. Though such special tableSource probably may not be suitable to be merged into Flink
master, in the future there could be other TableSourceTables which can support different join predicates depending on their different physical storage characteristics.

**Comment by fhueske** [comment]
The difference between failing during validation and translation is the code line that throws the exception. A check during validation causes an exception on the line that is responsible for the error. A check during translation throws an exception much later in the user program.

Regarding the check in the rule: IMO, each rule must ensure that is does not translate the query into an invalid plan. It make sense though, to have an additional safety check in the `DataSetJoin`. However, this check should be done in the constructor and not in the `translateToPlan()` method (see FLINK-5547) to fail fast if a rule tries to create an invalid plan.

**Comment by lincoln-lil** [comment]
@fhueske FLINK-5547 make sense to me,  I'll update this pr today.

**Comment by fhueske** [comment]
Great thanks!

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/sql/JoinITCase.scala [hunk 1]
```diff
@@ -83,6 +83,27 @@ class JoinITCase(
     val env = ExecutionEnvironment.getExecutionEnvironment
     val tEnv = TableEnvironment.getTableEnvironment(env, config)
 
+    val sqlQuery = "SELECT c, g FROM Table3, Table5 WHERE b = e AND a < 6"
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/sql/JoinITCase.scala]: I think this test covers the same case as the one above (`testJoinWithFilter()`).

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/table/JoinITCase.scala [hunk 2]
```diff
@@ -234,7 +249,38 @@ class JoinITCase(
...
+
+    val ds1 = CollectionDataSets.get3TupleDataSet(env).toTable(tEnv, 'a, 'b, 'c)
+    val ds2 = CollectionDataSets.get5TupleDataSet(env).toTable(tEnv, 'd, 'e, 'f, 'g, 'h)
+
+    val joinT = ds1.fullOuterJoin(ds2, 'a === 'd && 'b < 'h).select('c, 'g)
+
+    val results = joinT.toDataSet[Row].collect()
+  }
+
+
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/table/JoinITCase.scala]: Add tests to check that local predicates are not allowed in outer joins as well.

**Comment by lincoln-lil** [comment]
@fhueske thanks for your review, I'll update this pr according to your comments.

**Comment by lincoln-lil** [comment]
@fhueske when I add more test cases for "outer joins" with local predicates, I found different limitations between tableAPI and SQL, e.g., "SELECT c, g FROM Table3 left outer join Table5 on a = d AND e < 3" is valid, but dsFromTuple3.leftOuterJoin(dsFromTuple5, 'a === 'd && 'e < 3) is invalid, because validation in Join node sees original join predicates and validates simply, while in DataSetJoinRule the join predicates could be pruned according to the equivalence in relational algebra. 
I have a question, should we keep the tableAPI consistent with SQL? If so, the validation in tableAPI will be much more complex, and it reminds me the previous discussion about validation either in one or two places.  Or just maintain the current implementation unchanged, because we will add support for both FLINK-5498 and FLINK-5511 soon, then these differences will be eliminated.
What do you think ?

**Comment by fhueske** [comment]
I think it is fine to keep this small difference for now. With FLINK-5498 and FLINK-5511 these will be gone.

**Comment by fhueske** [comment]
Thanks for the update @lincoln-lil!
+1 to merge

**Pull request closed without merging**